core/store/src/archive/cold_storage.rs

```diff
@@ -399,23 +423,45 @@ pub fn test_get_store_initial_writes(column: DBCol) -> u64 {
 /// So, for every KeyType we need to capture all the keys that are related to that block.
 /// For BlockHash it is just one key -- block hash of that height.
 /// But for TransactionHash, for example, it is all of the tx hashes in that block.
+///
+/// Although `tracked_shards` should be sufficient to determine which shards matter,
+/// we process all shards as a precaution to ensure robustness and avoid
+/// missing data in case `tracked_shards` are incomplete due to bugs or future
+/// changes. This redundancy is acceptable because only the State column is
+/// performance- and size-sensitive, and it is handled separately in `copy_state_from_store`.
 fn get_keys_from_store(
     store: &Store,
     shard_layout: &ShardLayout,
+    tracked_shards: &Vec<ShardUId>,
     height_key: &[u8],
     block_hash_key: &[u8],
 ) -> io::Result<HashMap<DBKeyType, Vec<StoreKey>>> {
     let mut key_type_to_keys = HashMap::new();
+    let tracked_shards: HashSet<ShardId> =
+        tracked_shards.iter().map(|shard_uid| shard_uid.shard_id()).collect();
 
     let block: Block = store.get_ser_or_err_for_cold(DBCol::Block, &block_hash_key)?;
-    let chunks = block
-        .chunks()
-        .iter_deprecated()
-        .map(|chunk_header| {
-            store.get_ser_or_err_for_cold(DBCol::Chunks, chunk_header.chunk_hash().as_bytes())
-        })
-        .collect::<io::Result<Vec<ShardChunk>>>()?;
-
+    let mut chunk_hashes = vec![];
+    let mut chunks = vec![];
+    for chunk_header in block.chunks().iter_deprecated() {
+        let chunk_hash = chunk_header.chunk_hash();
+        chunk_hashes.push(chunk_hash.clone());
+        let chunk: Option<ShardChunk> =
+            store.get_ser_for_cold(DBCol::Chunks, chunk_hash.as_bytes())?;
+        let shard_id = chunk_header.shard_id();
```

Comment:
In case of an old chunk and resharding this may be shard id from previous shard layout. I won't mind if you add a todo here and merge as is for now but let's follow up and make it right. I don't know what 'right' means btw.

Reply:
Both `tracked_shards` and `shard_layout` are compatible with the `chunk_hash`' height.
https://github.com/near/nearcore/pull/13374/files#diff-89e9b4e52cb2022a47f22ea4cdbf50e00ca3bcf8b3d76c53c4d68141819bb72fR144-R146